core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -966,6 +966,66 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     go(None, this).stream
   }
 
+  /**
+    * Divide this streams into groups of elements received within a time window,
+    * or limited by the number of the elements, whichever happens first.
+    * Empty groups, which can occur if no elements can be pulled from upstream
+    * in a given time window, will not be emitted.
+    *
+    * Note: a time window starts each time downstream pulls.
+    */
+  def groupWithin[F2[x] >: F[x]](n: Int, d: FiniteDuration)(
+      implicit timer: Timer[F2],
+      F: Concurrent[F2]): Stream[F2, Chunk[O]] =
+    Stream
+      .eval(async.mutable.Queue.synchronousNoneTerminated[F2, Either[Token, Chunk[O]]])
+      .flatMap { q =>
+        def startTimeout: Stream[F2, Token] =
+          Stream.eval {
+            F.delay(new Token).flatTap { t =>
+              F.start { timer.sleep(d) *> q.enqueue1(t.asLeft.some) }
```

Comment:
Might not be necessary, especially for small `d`, but we might want to cancel this fiber on stream cancelation.

Reply:
good point, I can probably do this now?

```scala
Stream.eval(F.delay(new Token)).flatTap(t => Stream.supervise(timer.sleep(d) *> q.enqueue1(...)).void
```

Reply:
Yep, that looks like it would do it.

Reply:
I feel better about having `supervise` now 😀

Reply:
:)

I was actually missing something that `start`ed at the stream level (even though I wasn't coming from the pov of finalising `cancel`), so that's good to have